File: Codes/julia/utils/plot_settings.jl
# Updated on: 2025-05-12

## Color Set up
presentation_black = RGBf(0, 0, 0)
presentation_white = RGBf(1, 1, 1)
presentation_grey = RGBf(169/225, 169/225, 169/225)
presentation_red = RGBf(172/225, 68/225, 37/225)
presentation_green = RGBf(79/255, 121/255, 66/255)
presentation_blue = RGBf(15/255, 82/255, 186/255)
presentation_yellow = RGBf(186/255, 137/255, 104/255)
presentation_purple = RGBf(120/255, 81/255, 169/255)
presentation_teal = RGBf(52/255, 152/255, 164/255)
presentation_pink = RGBf(199/255, 94/255, 154/255)
presentation_orange = RGBf(217/255, 123/255, 42/255)

# Main Theme
main_plot_theme = Theme(
    size = (900, 600),
    fontsize = 24,
    fonts = (; regular = "Helvetica"),
    xlabelpadding = 20, ylabelpadding = 20,
)

set_theme!(main_plot_theme)

# Theme for black background
white_no_background_theme = Theme(
    size = (900, 600),
    fontsize = 24,
    fonts = (; regular = "Helvetica"),
    backgroundcolor = :transparent,
    Axis = (
        backgroundcolor = :transparent,
        xlabelcolor = presentation_white, ylabelcolor = presentation_white,
        xticklabelcolor = presentation_white, yticklabelcolor = presentation_white,
        ytickcolor = presentation_white, xtickcolor = presentation_white, 
        leftspinecolor = presentation_white, rightspinecolor = presentation_white,
        bottomspinecolor = presentation_white, topspinecolor = presentation_white,
        ygridcolor = presentation_white, xgridcolor = presentation_white,
    ),
    Legend = (
        backgroundcolor = :transparent,
        labelcolor = presentation_white,
        titlecolor = presentation_white,
        ticklabelcolor = presentation_white,
        tickcolor = presentation_white,
    ),
)

# Create a continuous colormap based on presentation colors
presentation_cmap_all = cgrad([presentation_blue, presentation_green, presentation_yellow, presentation_red,presentation_purple, presentation_teal, presentation_pink, presentation_orange], 256)

# Create a categorical colormap with n colors derived from presentation colors
function presentation_gen_colors(cmap, n=10)
    # Sample evenly spaced indices from the colormap
    indices = round.(Int, range(1, length(cmap), length=n))
    return cmap[indices]
end

# Crate quantile labels
function quant_labels(cut_off_list, range_q = false)
    quant_labels = []
    if range_q
        for i in 1:size(cut_off_list,1)-1
            lower_bound = @sprintf("%.2f", cut_off_list[i])
            upper_bound = @sprintf("%.2f", cut_off_list[i+1])
            push!(quant_labels, lower_bound*"-" *upper_bound)
        end   
    else 
        for i in 1:size(cut_off_list,1)
            lower_bound = @sprintf("%.2f", cut_off_list[i])
            push!(quant_labels, lower_bound)
        end
    end
    return quant_labels
end

# Define a moving average function
function moving_average(data, window_size)
    result = similar(data)
    n = length(data)
    for i in 1:n
        window_start = max(1, i - floor(Int, window_size/2))
        window_end = min(n, i + floor(Int, window_size/2))
        result[i] = mean(data[window_start:window_end])
    end
    return result
end

File: Project/analysis/routines/utils/plot_settings.py
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors
import numpy as np

# ==========================================
# 1. Color Set up
# ==========================================
presentation_black  = (0.0, 0.0, 0.0)
presentation_white  = (1.0, 1.0, 1.0)
presentation_grey   = (169/255, 169/255, 169/255)
presentation_red    = (172/255, 68/255, 37/255)
presentation_green  = (79/255, 121/255, 66/255)
presentation_blue   = (15/255, 82/255, 186/255)
presentation_yellow = (186/255, 137/255, 104/255)
presentation_purple = (120/255, 81/255, 169/255)
presentation_teal   = (52/255, 152/255, 164/255)
presentation_pink   = (199/255, 94/255, 154/255)
presentation_orange = (217/255, 123/255, 42/255)

# ==========================================
# 1. Makie-Style Theme Settings
# ==========================================
makie_style_theme = {
    # Figure and Axes
    'figure.figsize': (9, 6),
    'figure.facecolor': 'white',
    'axes.facecolor': 'white',
    
    # Spines (The bounding box)
    'axes.edgecolor': 'black',
    'axes.linewidth': 0.8,
    'axes.spines.top': True,
    'axes.spines.right': True,
    
    # Grid
    'axes.grid': True,
    'axes.axisbelow': True,        # Ensure grid is behind the points
    'grid.color': '#E5E5E5',       # Very light grey
    'grid.linewidth': 0.8,
    'grid.linestyle': '-',
    
    # Typography (Mimicking the academic/LaTeX look)
    'font.family': 'sans-serif',
    'font.sans-serif': ['Helvetica', 'Arial', 'DejaVu Sans'],
    'mathtext.fontset': 'dejavusans',      # DejaVu Sans for math symbols
    'font.size': 20,
    'axes.labelsize': 18,
    'axes.labelpad': 12.0,
    
    # Ticks
    'xtick.labelsize': 18,
    'ytick.labelsize': 18,
    'xtick.major.size': 5,         # Length of tick marks
    'ytick.major.size': 5,
    'xtick.direction': 'out',
    'ytick.direction': 'out',
}

# Apply the theme
plt.rcParams.update(makie_style_theme)

# Theme for black background
white_no_background_theme = {
    **makie_style_theme,                    # Inherit base settings
    'figure.facecolor': 'none',           # Transparent background
    'axes.facecolor': 'none',
    'savefig.facecolor': 'none',          # Transparent on save
    'text.color': presentation_white,
    'axes.labelcolor': presentation_white,
    'axes.edgecolor': presentation_white, # Spines
    'xtick.color': presentation_white,
    'ytick.color': presentation_white,
    'grid.color': presentation_white,
    'legend.facecolor': 'none',           # Transparent legend background
    'legend.edgecolor': 'none',
    'legend.labelcolor': presentation_white,
}

# Helper functions to switch themes easily
def set_main_theme():
    plt.rcParams.update(plt.rcParamsDefault)
    plt.rcParams.update(makie_style_theme)

def set_white_no_bg_theme():
    plt.rcParams.update(white_no_background_theme)

# ==========================================
# 3. Colormaps
# ==========================================
# Create a continuous colormap
color_list = [
    presentation_blue, presentation_green, presentation_yellow, 
    presentation_red, presentation_purple, presentation_teal, 
    presentation_pink, presentation_orange
]
presentation_cmap_all = mcolors.LinearSegmentedColormap.from_list(
    'presentation_cmap', color_list, N=256
)

# Create a categorical colormap
def presentation_gen_colors(cmap, n=10):
    """Samples evenly spaced indices from the colormap."""
    return [cmap(i) for i in np.linspace(0, 1, n)]

# ==========================================
# 4. Utility Functions
# ==========================================
def quant_labels(cut_off_list, range_q=False):
    """Creates quantile string labels."""
    labels = []
    if range_q:
        for i in range(len(cut_off_list) - 1):
            lower_bound = f"{cut_off_list[i]:.2f}"
            upper_bound = f"{cut_off_list[i+1]:.2f}"
            labels.append(f"{lower_bound}-{upper_bound}")
    else:
        for val in cut_off_list:
            labels.append(f"{val:.2f}")
    return labels

def moving_average(data, window_size):
    """
    Computes a centered moving average. Shrinks the window at the edges 
    to match the exact logic of the Julia implementation.
    """
    data = np.asarray(data)
    result = np.empty_like(data, dtype=float)
    n = len(data)
    half_window = window_size // 2
    
    for i in range(n):
        # Python uses 0-based indexing. Max/min bounds ensure we don't index out of bounds.
        window_start = max(0, i - half_window)
        window_end = min(n, i + half_window + 1) # +1 because Python slicing is exclusive
        result[i] = np.mean(data[window_start:window_end])
        
    return result